### DISCUSSION TITLE
Failed to use CliWrap to run a docker command

### DISCUSSION THREAD

**Discussion by citystrawman** [Tyrrrz/CliWrap#246] [Q&A]
Hi, I am trying to use CliWrap to run a docker command.  Actually it is a docker wrapper named fmriprep-docker(it is based upon fmri-prep, and fmriprep-docker is a Python script that operates the Docker Engine seamlessly, see [this ](https://fmriprep.org/en/1.5.4/installation.html#the-fmriprep-docker-wrapper)), when I execute it on powershell, it could run:
![企业微信截图_17185160625904](https://github.com/Tyrrrz/CliWrap/assets/9004748/e16dbfb1-84e5-481c-a98c-793b6b2d4e7a)
(the errors are because it requires some parameters, I do not use parameters for simplicity, but you can see it is able to run)
Then, I tried to implement it by CliWrap:
```
string workingDir = "C:\\Users\\PC\\AppData\\Roaming\\Python\\Python312\\Scripts";
string frmiprepDockerExe = Path.Combine(workingDir, "fmriprep-docker.exe");
var result = await Cli.Wrap(frmiprepDockerExe)
    .ExecuteBufferedAsync(cts.Token);
```
Howerver, when I run this code, it throws errors as follows:
![企业微信截图_17185164687426](https://github.com/Tyrrrz/CliWrap/assets/9004748/090281c5-6c9d-4a39-a5ca-4b0b8032ba90)

I searched docker though the discussion panel and I found one says that it may be related to TTY, but I am not quite clear from that post. I appreciate if anyone could help me solve this issue. Thank you!

**Comment by Tyrrrz** [comment]
99% chance it's the same issue as here: #190

**Reply by citystrawman** [reply]
Hi, Do you mean that I need to add `"."` at the end of argument? I changed my code as 
```
var result = await Cli.Wrap(frmiprepDockerExe)
    .WithArguments(new []{ "."}, false)
    .ExecuteBufferedAsync(cts.Token);
```
just like the way you did in #190, but still got the same result. 

Additionally, I tried my previous cliwrap code (they worked without any errors)  and they also showed the same error ("is not found!"), I think it may relate to some fundamental layers (probably some fundamental settings are changed in my powershell or wsl?)

**Reply by Tyrrrz** [reply]
Try running `echo "foo bar" | fmriprep-docker` in a terminal and see if it produces a similar error?

**Reply by citystrawman** [reply]
> Try running `echo "foo bar" | fmriprep-docker` in a terminal and see if it produces a similar error?
```
~$ echo "foo bar"|fmriprep-docker
RUNNING: docker run --rm -e DOCKER_VERSION_8395080871=26.1.3 -it -v /home/super_brain:/data:ro -v /home/super_brain:/out nipreps/fmriprep:23.2.3 /data /out participant
the input device is not a TTY
fMRIPrep: Please report errors to https://github.com/nipreps/fmriprep/issues
```
not sure if this is similar to #190, but  frmiprep-docker indeed runs.

**Reply by citystrawman** [reply]
> Try running `echo "foo bar" | fmriprep-docker` in a terminal and see if it produces a similar error?

BTW I tried this in WSL. in powershell, when I tried bash.exe -c "fmriprep-docker", it said `/bin/bash: fmriprep-docker: command not found`

**Reply by Tyrrrz** [reply]
Does it work if you `Cli.Wrap("pwsh").WithArguments(["-command", "fmriprep-docker"])` (or whatever is the right input to run your app via `pwsh`?

**Reply by citystrawman** [reply]
> Does it work if you `Cli.Wrap("pwsh").WithArguments(["-command", "fmriprep-docker"])` (or whatever is the right input to run your app via `pwsh`?

Currently I can run the code perfectly through this:
![企业微信截图_17186401151463](https://github.com/Tyrrrz/CliWrap/assets/9004748/6cb7d55f-d0b2-4d53-a0f5-a363a4518d25)
and I use CliWrap to simulate this:

```
var result = await Cli.Wrap(bashPath)
    .WithArguments(args => args
    .Add("-c")
    .Add(q))
    .ExecuteBufferedAsync(cts.Token);
```
where bashPath = "C:\\Windows\\System32\\bash.exe"
and q is exactly the string from the above screenshot:
![image](https://github.com/Tyrrrz/CliWrap/assets/9004748/60f19b7f-221f-4a83-a93d-9c28ef639397)

Then I got exceptions: 
![企业微信截图_17186404721250](https://github.com/Tyrrrz/CliWrap/assets/9004748/7d7138f5-4265-4695-a27b-5742e9d7b4a0)

Then I tried another code:
```
var result = await Cli.Wrap(pwshPath)
    .WithArguments(args => args
    .Add("bash.exe")
    .Add("-c")
    .Add(q))
    .ExecuteBufferedAsync(cts.Token);
```
but here it returns result immediately and the result is successful. However , this command will actually execute about 50 minutes . I checked the output directory and it seems that it is producing the results, but I have to wait 50 minutes to check if it really doing OK.  

BTW this discussion is too messy, some of my previous posts(for example, the "is not found" error) are due to my own fault. So may I close it and start another discussion?

**Reply by Tyrrrz** [reply]
The "input device is not a TTY" does seem like a standard input issue. It can happen when the underlying program does not check whether the standard input has been redirected before performing certain user interactions. With CliWrap all standard streams are always redirected, which is what makes the behavior different from running the command in a shell. If the underlying program (`fmriprep` in this case) doesn't provide a way to optionally disregard standard input (as it was possible with `ripgrep` in the linked issue), it might not be possible to properly wrap it. There is an open issue to support such scenarios (#225) but it's quite complex and hasn't been worked on.

> BTW this discussion is too messy, some of my previous posts(for example, the "is not found" error) are due to my own fault. So may I close it and start another discussion?

Sure.

**Reply by citystrawman** [reply]
Hi Sir, I really appreciate your help and I want to sponsor you, however the only effective payment way that I can use is alipay. Is it possible for you to receive donation by alipay?

**Reply by Tyrrrz** [reply]
Thank you. Unfortunately I can't use Alipay :(

The only donation options I have are:
- [GitHub Sponsors](https://github.com/sponsors/Tyrrrz)
- [Patreon](https://patreon.com/Tyrrrz)
- [Buy Me a Coffee](https://buymeacoffee.com/Tyrrrz)

If they don't work for you, don't worry about it :)

**Reply by citystrawman** [reply]
> Thank you. Unfortunately I can't use Alipay :(
> 
> The only donation options I have are:
> 
> * [GitHub Sponsors](https://github.com/sponsors/Tyrrrz)
> * [Patreon](https://patreon.com/Tyrrrz)
> * [Buy Me a Coffee](https://buymeacoffee.com/Tyrrrz)
> 
> If they don't work for you, don't worry about it :)

Oh sorry, I tried all 3 methods using my bank card but none of them work (I have no credit card but only bank card which support union pay). Wish someday you can support alipay or union pay!
